From eb8025bcaf25d9ddf2ca6a7015a13b179b239c9b Mon Sep 17 00:00:00 2001
From: Tom Hennigan <tomhennigan@google.com>
Date: Thu, 18 Nov 2021 10:16:49 -0800
Subject: [PATCH] Maintain per-stream runner cache.

After cl/410164595 users reported seeing:

    CudnnLegacyConvRunner cached across multiple StreamExecutors.

Multiple models are affected at least one using the spmd partitioner (via
jax.pjit) and others using normal data parallelism (using jax.pmap).

PiperOrigin-RevId: 410837882
Change-Id: Idbd0b96a6c3aa2a6a333ffd6e95982e36911b674
---
 .../xla/service/gpu/convolution_thunk.cc       | 18 +++++++++++++++---
 .../xla/service/gpu/convolution_thunk.h        |  8 +++++++-
 2 files changed, 22 insertions(+), 4 deletions(-)

diff --git a/tensorflow/compiler/xla/service/gpu/convolution_thunk.cc b/tensorflow/compiler/xla/service/gpu/convolution_thunk.cc
index fa6e9d473eee75..b82c7f5683a0f7 100644
--- a/tensorflow/compiler/xla/service/gpu/convolution_thunk.cc
+++ b/tensorflow/compiler/xla/service/gpu/convolution_thunk.cc
@@ -15,6 +15,7 @@ limitations under the License.
 
 #include "tensorflow/compiler/xla/service/gpu/convolution_thunk.h"
 
+#include <memory>
 #include <string>
 
 #include "absl/strings/str_cat.h"
@@ -37,8 +38,19 @@ ConvolutionThunk::ConvolutionThunk(
       operand_buffers_(std::move(operand_slices)),
       result_buffer_(result_slice),
       scratch_buffer_(scratch_slice),
-      config_(std::move(config)),
-      runner_cache_(config_) {}
+      config_(std::move(config)) {}
+
+MaybeFusedConvRunner& ConvolutionThunk::GetOrCreateRunner(
+    const stream_executor::Stream* stream) {
+  tensorflow::mutex_lock lock(mu_);
+  auto it = runner_cache_.find(stream);
+  if (it == runner_cache_.end()) {
+    it = runner_cache_
+             .insert({stream, std::make_unique<MaybeFusedConvRunner>(config_)})
+             .first;
+  }
+  return *it->second;
+}
 
 Status ConvolutionThunk::ExecuteOnStream(const ExecuteParams& params) {
   const auto& buffer_allocations = *params.buffer_allocations;
@@ -55,7 +67,7 @@ Status ConvolutionThunk::ExecuteOnStream(const ExecuteParams& params) {
       buffer_allocations.GetDeviceAddress(scratch_buffer_);
 
   RunConvOptions opts;
-  opts.runner_cache = &runner_cache_;
+  opts.runner_cache = &GetOrCreateRunner(params.stream);
 
   TF_RETURN_IF_ERROR(RunGpuConv(config_, absl::MakeSpan(operand_se_buffers),
                                 result_buffer, scratch, params.stream, opts));
diff --git a/tensorflow/compiler/xla/service/gpu/convolution_thunk.h b/tensorflow/compiler/xla/service/gpu/convolution_thunk.h
index 3c695f6b824902..99267530b5253d 100644
--- a/tensorflow/compiler/xla/service/gpu/convolution_thunk.h
+++ b/tensorflow/compiler/xla/service/gpu/convolution_thunk.h
@@ -16,6 +16,7 @@ limitations under the License.
 #ifndef TENSORFLOW_COMPILER_XLA_SERVICE_GPU_CONVOLUTION_THUNK_H_
 #define TENSORFLOW_COMPILER_XLA_SERVICE_GPU_CONVOLUTION_THUNK_H_
 
+#include "absl/container/flat_hash_map.h"
 #include "absl/types/optional.h"
 #include "tensorflow/compiler/xla/service/buffer_assignment.h"
 #include "tensorflow/compiler/xla/service/gpu/buffer_allocations.h"
@@ -55,10 +56,15 @@ class ConvolutionThunk : public Thunk {
   std::vector<BufferAllocation::Slice> operand_buffers_;
   BufferAllocation::Slice result_buffer_;
   BufferAllocation::Slice scratch_buffer_;
+  MaybeFusedConvRunner& GetOrCreateRunner(
+      const stream_executor::Stream* stream);
 
   // Convolution config
   const GpuConvConfig config_;
-  MaybeFusedConvRunner runner_cache_;
+  tensorflow::mutex mu_;
+  absl::flat_hash_map<const stream_executor::Stream*,
+                      std::unique_ptr<MaybeFusedConvRunner>>
+      runner_cache_ TF_GUARDED_BY(mu_);
 };
 
 }  // namespace gpu
